### DISCUSSION TITLE
Chord and result backend questions

### DISCUSSION THREAD

**Discussion by wrkhenddher** [celery/celery#7165] [Q&A]
Hello all:

While trying some multi-task synchronization, I *discovered* some nuances. 

```python
@app.task()
def subtask(...):
   ...
@app.task()
def subtasks_completed(...):

subtasks = [subtask.si(*st_args) for st_args in args]
wf = group(*subtasks) | subtasks_completed.si(*args)
wf.apply_async()
```

I.e., I am spawning a bunch of subtasks, each one with its own immutable set of arguments `st_args`, followed by a unique task `subtasks_completed`. None of them returns any result whatsoever. i.e. spawn and bunch of subtasks and join them to a final subtasks_completed.

My questions:

- Is a backend result mandatory when `chord` is used? 

Both celery 3.x and 4.x seem to fail to execute chords 'til completion unless a result backend is specified. Even when none of the tasks within the chord returns results. IOW, when I didn't have Redis configured as backend, `subtasks_completed` sometimes would not be executed (randomly). It wasn't clear if the subtasks_completed task was ever scheduled as neither RabbitMQ nor Celery "saw" it. Ended up setting `CELERY_IGNORE_RESULT = False` globally and Redis as results backend to prevent the random "disappearance" of the `subtasks_completed`.

- When Redis is used as result backend, `KEYS`s `celery-task*` are created but they are never removed.  [Documentation](http://docs.celeryproject.org/en/latest/getting-started/first-steps-with-celery.html#keeping-results) states that results must be "discarded" (`get()` or `forget()`) but does that apply to tasks that are NOT returning any results too?

- If each individual task `forget()`s its own result, will the `chord` be affected?
```python
@app.task(...)
def subtask(...):
    ...
    r = self.AsyncResult(task_id=self.request.id)
    r.forget()
```

```
software -> celery:4.2.0 (windowlicker) kombu:4.2.1 py:3.6.3
            billiard:3.5.0.3 py-amqp:2.2.2
platform -> system:Darwin arch:64bit imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:amqp results:redis://localhost:6379/0

BROKER_URL: 'amqp://guest:********@localhost:5672//'
CELERY_RESULT_BACKEND: 'redis://localhost:6379/0'
CELERY_TASK_SERIALIZER: 'json'
CELERY_ACCEPT_CONTENT: ['json']
CELERY_RESULT_SERIALIZER: 'json'
CELERY_TIMEZONE: 'US/Central'
CELERY_ENABLE_UTC: True
CELERY_IGNORE_RESULT: False
CELERYD_PREFETCH_MULTIPLIER: 1
CELERY_ACKS_LATE: True
CELERYD_LOG_COLOR: True
CELERYD_LOG_FORMAT: ('[%(asctime)s: %(levelname)s/%(processName)s] '
 '%(name)s %(message)s')
CELERYD_TASK_LOG_FORMAT: ('[%(asctime)s: %(levelname)s/%(processName)s] '
 '[%(task_name)s(%(task_id)s)] %(name)s '
 '%(message)s')
```

**Comment by wrkhenddher** [comment]
Tried `self.AsyncResult(task_id=self.request.id).forget()` at the end of each `@app.task` and Redis key/values were never purged either (#4660) 

They seem to be purged once `app.conf.CELERY_TASK_RESULT_EXPIRES` elapses.

**Comment by georgepsarakis** [comment]
@wjehenddher some answers to your questions:
- `chord` workflows require a result backend to be configured, I believe it is also stated in the documentation.
- `forget` will not work from inside the task, because the result is stored, after the task completes.

**Comment by wrkhenddher** [comment]
Thank you @georgepsarakis
